File: app/src/main/java/com/capstone/project_niyakneyak/ui/login/activity/LoginActivity.java
package com.capstone.project_niyakneyak.login.activity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.capstone.project_niyakneyak.databinding.ActivityLoginBinding;
import com.capstone.project_niyakneyak.ui.login.etc.LoginFormState;
import com.capstone.project_niyakneyak.ui.login.etc.LoginResult;
import com.capstone.project_niyakneyak.ui.login.viewmodel.LoginViewModel;
import com.capstone.project_niyakneyak.ui.login.viewmodel.LoginViewModelFactory;
import com.capstone.project_niyakneyak.ui.main.activity.MainActivity;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;

public class LoginActivity extends AppCompatActivity {
    private LoginViewModel loginViewModel;
    private ActivityLoginBinding binding;
    private static final String TAG = "GoogleActivity";
    private GoogleSignInClient mGoogleSignInClient;
    private FirebaseAuth mAuth;

    String default_web_client_id = "264335307721-6gi684ianko5vqjira4o4uqttlppl6ab.apps.googleusercontent.com";

    @Override
    public void onStart() {
        super.onStart();
        // Check if user is signed in (non-null) and update UI accordingly.
        FirebaseUser currentUser = mAuth.getCurrentUser();
        if(currentUser != null)
        {
            updateUI(currentUser);
        }

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(default_web_client_id)
                .requestEmail()
                .build();

        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        mAuth = FirebaseAuth.getInstance();

        binding = ActivityLoginBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        loginViewModel = new ViewModelProvider(this, new LoginViewModelFactory())
                .get(LoginViewModel.class);

        loginViewModel.getLoginFormState().observe(this, new Observer<LoginFormState>() {
            @Override
            public void onChanged(@Nullable LoginFormState loginFormState) {
                if (loginFormState == null) {
                    return;
                }
                binding.mainLoginPlain.setEnabled(loginFormState.isDataValid());
                if (loginFormState.getUsernameError() != null) {
                    binding.mainIdLayout.setErrorEnabled(true);
                    binding.mainIdLayout.setError(getString(loginFormState.getUsernameError()));
                }
                else binding.mainIdLayout.setErrorEnabled(false);

                if (loginFormState.getPasswordError() != null) {
                    binding.mainPasswordLayout.setErrorEnabled(true);
                    binding.mainPasswordLayout.setError(getString(loginFormState.getPasswordError()));
                }
                else binding.mainPasswordLayout.setErrorEnabled(false);
            }
        });

        loginViewModel.getLoginResult().observe(this, new Observer<LoginResult>() {
            @Override
            public void onChanged(@Nullable LoginResult loginResult) {
                if (loginResult == null) {
                    return;
                }
                binding.loadingBarPlain.setVisibility(View.GONE);
                binding.loadingBarGoogle.setVisibility(View.GONE);
                binding.loadingBarNaver.setVisibility(View.GONE);
                if (loginResult.getError() != null) {
                    showLoginFailed(loginResult.getError());
                }
                if (loginResult.getSuccess() != null) {
                    updateUiWithUser();
                }
            }
        });

        TextWatcher afterTextChanged = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }
            @Override
            public void afterTextChanged(Editable s) {
                loginViewModel.loginDataChanged(binding.mainIdEditText.getText().toString(),
                        binding.mainPwEditText.getText().toString());
            }
        };

        binding.mainIdEditText.addTextChangedListener(afterTextChanged);
        binding.mainPwEditText.addTextChangedListener(afterTextChanged);
        binding.mainPwEditText.setOnEditorActionListener((v, actionId, event) -> {
            if(actionId == EditorInfo.IME_ACTION_DONE){
                loginViewModel.login(binding.mainIdEditText.getText().toString(),
                        binding.mainPwEditText.getText().toString());
            }
            return false;
        });

        binding.mainLoginPlain.setOnClickListener(v -> {
            binding.loadingBarPlain.setVisibility(View.VISIBLE);
            loginViewModel.login(binding.mainIdEditText.getText().toString(),
                    binding.mainPwEditText.getText().toString());
        });

        binding.mainLoginPlain.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                binding.loadingBarPlain.setVisibility(View.VISIBLE);
                String email = binding.mainIdEditText.getText().toString();
                String password = binding.mainPwEditText.getText().toString();

                mAuth.signInWithEmailAndPassword(email,password).addOnCompleteListener(LoginActivity.this, new OnCompleteListener<AuthResult>()
                {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            // Sign in success, update UI with the signed-in user's information
                            Intent intent = new Intent (LoginActivity.this, MainActivity.class);
                            startActivity(intent);
                            Toast.makeText(LoginActivity.this, "로그인 성공", Toast.LENGTH_SHORT).show();
                            finish();
                        } else {
                            // If sign in fails, display a message to the user.
                            Toast.makeText(LoginActivity.this, "로그인 실패", Toast.LENGTH_SHORT).show();

                        }
                    }
                });
            }
        });

        binding.mainLoginGoogle.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                binding.loadingBarGoogle.setVisibility(View.VISIBLE);
                signIn();
            }
        });

        binding.mainRegisterBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //회원가입 화면으로 이동
                Intent intent = new Intent(LoginActivity.this, RegisterActivity.class); //현재 acitivity, 이동할 activity
                startActivity(intent);//액티비티 이동
            }
        });

    }

    private void updateUiWithUser() {
        Intent intent = new Intent(LoginActivity.this, MainActivity.class);
        intent.putExtra("USER_ID", binding.mainIdEditText.getText().toString());
        intent.putExtra("USER_PW", binding.mainPwEditText.getText().toString());
        startActivityIfNeeded(intent, 0);
    }

    private void showLoginFailed(@StringRes Integer errorString) {
        Toast.makeText(getApplicationContext(), errorString, Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(resultCode == RESULT_OK){
            loginViewModel.logout(data.getStringExtra("USER_ID"), data.getStringExtra("USER_PW"));
        }
    }

    private void firebaseAuthWithGoogle(String idToken) {
        AuthCredential credential = GoogleAuthProvider.getCredential(idToken, null);
        mAuth.signInWithCredential(credential)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            // Sign in success, update UI with the signed-in user's information
                            Log.d(TAG, "signInWithCredential:success");
                            FirebaseUser user = mAuth.getCurrentUser();
                            updateUI(user);
                        } else {
                            // If sign in fails, display a message to the user.
                            Log.w(TAG, "signInWithCredential:failure", task.getException());
                            updateUI(null);
                        }
                    }
                });
    }

    private void updateUI(FirebaseUser user) {
        if (user != null) {
            // User is signed in
            String userId = user.getUid();
            String userEmail = user.getEmail();
            // Perform actions based on the signed-in user, such as navigating to the main activity

            // Example: Navigate to MainActivity
            Intent intent = new Intent(LoginActivity.this, MainActivity.class);
            intent.putExtra("USER_ID", userId);
            intent.putExtra("USER_EMAIL", userEmail);
            startActivity(intent);
            finish(); // Optional: Close the login activity to prevent going back to it using the back button
        } else {
            signOut();
            Intent intent = new Intent(getApplicationContext(), LoginActivity.class);
            startActivity(intent);
            finish();
        }
    }

    ActivityResultLauncher<Intent> launcher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK) {
                    Intent data = result.getData();
                    handleSignInResult(data);
                }
            });

    // 사용할 곳에서 다음과 같이 사용
    public void signIn() {
        Intent signInIntent = mGoogleSignInClient.getSignInIntent();
        launcher.launch(signInIntent);
    }

    private void handleSignInResult(Intent data) {
        if (data != null) {
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            try {
                GoogleSignInAccount account = task.getResult(ApiException.class);
                Log.d(TAG, "firebaseAuthWithGoogle:" + account.getId());
                firebaseAuthWithGoogle(account.getIdToken());
            } catch (ApiException e) {
                Log.w(TAG, "Google sign in failed", e);
                updateUI(null);
            }
        }
    }

    public void signOut(){
        Log.d(TAG, "signOut: " + mAuth.getCurrentUser().getEmail());
        mAuth.signOut();
    }
}

File: app/src/main/java/com/capstone/project_niyakneyak/login/activity/LoginActivity.kt
package com.capstone.project_niyakneyak.login.activity

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.SpannableString
import android.text.TextWatcher
import android.text.style.UnderlineSpan
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.capstone.project_niyakneyak.alarm.service.RescheduleAlarmService
import com.capstone.project_niyakneyak.data.alarm_model.Alarm
import com.capstone.project_niyakneyak.data.alarm_valid_model.AlarmV
import com.capstone.project_niyakneyak.data.user_model.UserAccount
import com.capstone.project_niyakneyak.databinding.ActivityLoginBinding
import com.capstone.project_niyakneyak.login.etc.LoggedInUserView
import com.capstone.project_niyakneyak.login.etc.LoginResult
import com.capstone.project_niyakneyak.login.viewmodel.LoginViewModel
import com.capstone.project_niyakneyak.main.activity.AppSettingActivity.Companion.TAG
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.firebase.Firebase
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.auth
import com.google.firebase.firestore.Filter
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.firestore


class LoginActivity : AppCompatActivity() {
    private lateinit var firestore: FirebaseFirestore
    private lateinit var firebaseAuth: FirebaseAuth
    private lateinit var binding: ActivityLoginBinding
    private lateinit var loginViewModel: LoginViewModel

    private var gso: GoogleSignInOptions? = null
    private var mGoogleSignInClient: GoogleSignInClient? = null

    private val googleSignInLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            if (it.resultCode == RESULT_OK) {
                val data = it.data
                handleGoogleSignInResult(data)
            }
        }
    private val signUpLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()){
        if(it.resultCode == RESULT_OK){
            loginViewModel.loginResult.value = LoginResult(LoggedInUserView(firebaseAuth.currentUser!!.uid))
        }
    }

    companion object{
        private const val default_web_client_id = "264335307721-6gi684ianko5vqjira4o4uqttlppl6ab.apps.googleusercontent.com"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val requestToken = intent.getIntExtra("request_token", 0)

        // Initialize GoogleSignInOptions and FirebaseAuth
        gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestIdToken(default_web_client_id)
            .requestEmail()
            .build()
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso!!)

        firestore = Firebase.firestore
        firebaseAuth = Firebase.auth

        // Setting View by Binding
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Initialize LoginViewModel
        loginViewModel = ViewModelProvider(this)[LoginViewModel::class.java]

        // Observe LoginForm
        loginViewModel.loginFormState.observe(this, Observer { loginFormState ->
            if (loginFormState == null) {
                return@Observer
            }
            binding.mainLoginPlain.isEnabled = loginFormState.isDataValid
            if (loginFormState.usernameError != null) {
                binding.mainIdLayout.isErrorEnabled = true
                binding.mainIdLayout.error = getString(loginFormState.usernameError!!)
            } else binding.mainIdLayout.isErrorEnabled = false
            if (loginFormState.passwordError != null) {
                binding.mainPasswordLayout.isErrorEnabled = true
                binding.mainPasswordLayout.error = getString(loginFormState.passwordError!!)
            } else binding.mainPasswordLayout.isErrorEnabled = false
        })

        // Observe LoginResult
        loginViewModel.loginResult.observe(this, Observer { loginResult ->
            if (loginResult == null) return@Observer

            binding.loadingBarPlain.visibility = View.GONE
            binding.loadingBarGoogle.visibility = View.GONE
            if (loginResult.error != null) {
                showLoginFailed(loginResult.error!!)
            }
            if (loginResult.success != null) {
                firestore.collection(UserAccount.COLLECTION_ID).document(firebaseAuth.currentUser!!.uid)
                    .collection(Alarm.COLLECTION_ID).where(Filter.equalTo(Alarm.FIELD_IS_STARTED, true)).get()
                    .addOnSuccessListener {
                        if(requestToken == 1){
                            val intentService = Intent(applicationContext, RescheduleAlarmService::class.java)
                            applicationContext.startService(intentService)
                        } else{
                            val alarmV = AlarmV(alarmCode = 1, isStarted = true)
                            alarmV.scheduleAlarm(applicationContext)
                        }
                        setResult(RESULT_OK)
                        finish()
                    }.addOnFailureListener {
                        Log.w("RescheduleService", "Reschedule Failed: $it")
                        setResult(RESULT_OK)
                        finish()
                    }
            }
        })

        // Setting TextChangedListener in Id and Pw EditText
        val afterTextChanged: TextWatcher = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                loginViewModel.loginDataChanged(binding.mainIdEditText.text.toString(), binding.mainPwEditText.text.toString())
            }
        }
        binding.mainIdEditText.addTextChangedListener(afterTextChanged)
        binding.mainPwEditText.addTextChangedListener(afterTextChanged)
        binding.mainPwEditText.setOnEditorActionListener { _: TextView?, actionId: Int, _: KeyEvent? ->
            if (actionId == EditorInfo.IME_ACTION_DONE)
                binding.mainLoginPlain.callOnClick()
            false
        }

        // Setting onClickListener in each button
        // Login Process
        binding.mainLoginPlain.setOnClickListener {
            binding.loadingBarPlain.visibility = View.VISIBLE
            emailSignIn()
        }
        binding.mainLoginGoogle.setOnClickListener {
            binding.loadingBarGoogle.visibility = View.VISIBLE
            googleSignIn()
        }
        // Register Process
        binding.mainRegisterBtn.setOnClickListener {
            // 회원가입 화면으로 이동
            val intent = Intent(this, RegisterActivity::class.java)
            signUpLauncher.launch(intent)
        }

        binding.resetPassword.setOnClickListener {
            // 밑줄 추가
            val content = SpannableString("비밀번호를 잊어버렸습니다")
            content.setSpan(UnderlineSpan(), 0, content.length, 0)
            binding.resetPassword.text = content

            val intent = Intent(this, ResetPasswordActivity::class.java)
            startActivity(intent)
        }
    }

    // Handle Email SignIn Process
    private fun emailSignIn(){
        val email = binding.mainIdEditText.text.toString()
        val password = binding.mainPwEditText.text.toString()
        firebaseAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener(this) { task ->
            if (task.isSuccessful) {
                // Change LiveData about LoginResult as Success
                loginViewModel.loginResult.value = LoginResult(LoggedInUserView(firebaseAuth.currentUser!!.uid))
            } else {
                // Change LiveData about LoginResult as Failed
                loginViewModel.loginResult.value = LoginResult(task.exception)
            }
        }
    }


    // Handle Google SignIn Process
    private fun googleSignIn() {
        val signInIntent = mGoogleSignInClient!!.signInIntent
        googleSignInLauncher.launch(signInIntent)
    }

    private fun firebaseAuthWithGoogle(idToken: String?) {
        val credential = GoogleAuthProvider.getCredential(idToken, null)
        firebaseAuth.signInWithCredential(credential)
            .addOnCompleteListener(this) { task ->
                if (task.isSuccessful) {
                    //view모델을 통해서 바꿔서 loginResult를 바꿔줘야함
                    //observe해서 실시간으로 변화를 보면서 값이 변하면 그에 따라서, 절차를 수행

                    val isNewUser = task.result?.additionalUserInfo?.isNewUser == true  // 새 사용자 여부 확인
                    firebaseAuth.currentUser?.let { firebaseUser ->
                        if (isNewUser) {
                            // 새 사용자인 경우 Firestore에 사용자 정보 저장 및 GoogleRegisterActivity로 이동
                            saveUserToFirestore(firebaseUser)
                        } else {
                            // 기존 사용자인 경우 MainActivity로 이동
                            loginViewModel.loginResult.value = LoginResult(LoggedInUserView(firebaseUser.uid))
                        }
                    } ?: run {
                        Log.w(TAG, "Firebase User is null after sign-in success.")
                        showLoginFailed(Exception("Authentication succeeded but user is null"))
                    }
                } else {
                    // 로그인 실패 처리
                    task.exception?.let {
                        loginViewModel.loginResult.value = LoginResult(it)
                        showLoginFailed(it)
                    }
                }
            }
    }

    private fun saveUserToFirestore(user: FirebaseUser) {
        val userMap = hashMapOf("email" to (user.email ?: ""))
        firestore.collection("users").document(user.uid).set(userMap)
            .addOnSuccessListener {
                Log.d( "Firestore", "User successfully written!")
                Toast.makeText(this, "사용자 정보가 성공적으로 등록되었습니다.", Toast.LENGTH_SHORT).show()
                navigateToGoogleRegisterActivity(user)
            }
            .addOnFailureListener { e ->
                Log.w("Firestore", "Error writing document", e)
                Toast.makeText(this, "사용자 정보 등록 실패: ${e.localizedMessage}", Toast.LENGTH_SHORT).show()
            }
    }

    private fun navigateToGoogleRegisterActivity(firebaseUser: FirebaseUser) {
        Intent(this, GoogleRegisterActivity::class.java).apply {
            putExtra("uid", firebaseUser.uid)
            putExtra("email", firebaseUser.email ?: "")
            putExtra("name", firebaseUser.displayName ?: "")
            startActivity(this)
        }
        finish()
    }

    private fun handleGoogleSignInResult(data: Intent?) {
        if (data != null) {
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            try {
                val account = task.getResult(ApiException::class.java)
                firebaseAuthWithGoogle(account.idToken)
            } catch (e: ApiException) {
                loginViewModel.loginResult.value = LoginResult(e)
            }
        }
    }

    // Handle Login Failure
    private fun showLoginFailed(errorString: Exception) {
        Toast.makeText(this, errorString.toString(), Toast.LENGTH_SHORT).show()
    }

}